File: haskell/src/Day1.hs
module Day1
  ( runPart1,
    runPart2,
  )
where

import Util

input :: String
input = "../inputs/day1.txt"

readInput :: IO [Integer]
readInput =
  do
    fileContents <- readFile input
    fileContents
      |> lines
      |> map read
      |> pure

solvePart1 :: [Integer] -> String
solvePart1 numbers =
  numbers
    |> combinations 2
    |> filter (\comb -> sum comb == 2020)
    |> head
    |> product
    |> show

solvePart2 :: [Integer] -> String
solvePart2 numbers =
  numbers
    |> combinations 3
    |> filter (\comb -> sum comb == 2020)
    |> head
    |> product
    |> show

runPart1 :: IO ()
runPart1 = do
  numbers <- readInput
  putStrLn $ solvePart1 numbers

runPart2 :: IO ()
runPart2 = do
  numbers <- readInput
  putStrLn $ solvePart2 numbers

File: rust/src/aoc/day1.rs
use itertools::Itertools;
use std::fs;
use std::io;
use std::io::prelude::*;
use std::iter::Iterator;

const INPUT: &str = "../inputs/day1.txt";

fn read_input() -> impl Iterator<Item = usize> {
    let input = fs::File::open(INPUT).unwrap();
    let file_contents = io::BufReader::new(input);

    file_contents.lines().map(|s| s.unwrap().parse().unwrap())
}

#[allow(dead_code)]
pub fn run_part_1() {
    let res: Vec<usize> = read_input()
        .combinations(2)
        .find(|p| p.iter().sum::<usize>() == 2020)
        .unwrap();

    println!("{:?}", res.iter().product::<usize>());
}

#[allow(dead_code)]
pub fn run_part_2() {
    let res: Vec<usize> = read_input()
        .combinations(3)
        .find(|p| p.iter().sum::<usize>() == 2020)
        .unwrap();

    println!("{:?}", res.iter().product::<usize>());
}
